### PR TITLE
Install Helm

### PR DESCRIPTION
**Pull request by dkoshkin**
Fixes: #509 

Added new plan file options:
```
cluster...
features:
  package_manager:
    enabled: true
    provider: helm
```

Modified output:
```
Validating==========================================================================
Reading installation plan file "kismatic-cluster.yaml"                          [OK]
Validating installation plan file                                               [OK]
Validating SSH connectivity to nodes                                            [OK]

Generating Kubeconfig File==========================================================
Generated kubeconfig file in the "generated" directory                          [OK]

Configuring Certificates============================================================
Found valid key and certificate for node "etcd001"                              [OK]
Found valid key and certificate for node "master001"                            [OK]
Found valid key and certificate for node "worker001"                            [OK]
Found certificate for docker registry                                           [OK]
Found key and certificate for service accounts                                  [OK]
Found key and certificate for user "admin"                                      [OK]
Cluster certificates can be found in the "generated" directory                  [OK]

Installing Cluster==================================================================
...

Installing Helm on the Cluster======================================================
Backed up "/Users/dkoshkin/.helm" directory                                     [OK]
Installed Tiller (the helm server side component)                               [OK]
Create Required RBAC Role for Helm
- Running task:
  master001                                                                     [OK]
- Running task: Create new serviceaccount for tiller
  master001                                                                     [OK]
- Running task: Attach cluster-admin role to the service account tiller
  master001                                                                     [OK]
- Running task: Patch the tiller spec to use the new serviceaccount
  master001                                                                     [OK]
Create Required RBAC Role for Helm Finished

Running Smoke Test==================================================================
Smoke Test Master Node
- Running task:
  master001                                                                     [OK]
- Running task: copy Kuberang to node
  master001                                                                     [OK]
- Running task: run smoke test checks using Kuberang
  master001                                                                     [OK]
Smoke Test Master Node Finished

The cluster was installed successfully!

- To use the generated kubeconfig file with kubectl:
    * use "kubectl --kubeconfig generated/kubeconfig"
    * or copy the config file "cp generated/kubeconfig ~/.kube/config"
- To view the Kubernetes dashboard: "./kismatic dashboard"
- To SSH into a cluster node: "./kismatic ssh etcd|master|worker|storage|$node.host"
```

### CODE REVIEW

**Comment by eicnix** [comment]
Will this work with RBAC enabled cluster? AFAIK you don't create a serviceaccount and rolebinding for tiller. [Example](https://github.com/mgoodness/prometheus-operator/blob/d765e5651fa7acad51b236971d20cb32d9675a35/helm/prometheus-operator/README.md#rbac)

**Comment by dkoshkin** [comment]
Good question @eicnix, I will take a look at that, thanks.

File: pkg/cli/apply.go [hunk 1]
```diff
@@ -88,34 +89,62 @@ func (c *applyCmd) run() error {
...
+		// Run 'helm init'
+		if err := helm.Init(); err != nil {
+			util.PrettyPrintErr(c.out, "Installed Helm on the cluster")
+			return fmt.Errorf("error installing Helm on the cluster: %v", err)
+		}
+		util.PrettyPrintOk(c.out, "Installed Tiller (the helm server side component)")
+
+		// HelmRBAC will create a new role RBAC manually
+		// TODO remove when https://github.com/kubernetes/helm/issues/2224 gets fully fixed
+		if err := c.executor.RunPlay("_helm-rbac.yaml", plan); err != nil {
```
Comment (alexbrand) [inline pkg/cli/apply.go]: Would it be better to run this as part of the overall kubernetes play? We can have a new ansible var that will control it.

Comment (alexbrand) [inline pkg/cli/apply.go]: I guess the play waits until tiller is up and running. Hmmm.... This makes me wonder if calling `helm init` should be done from ansible as well.

File: pkg/install/execute.go [hunk 2]
```diff
@@ -681,21 +680,17 @@ func (ae *ansibleExecutor) buildClusterCatalog(p *Plan) (*ansible.ClusterCatalog
...
+		cc.DockerCAPath = filepath.Join(tlsDir, "ca.pem")
+		cc.DockerRegistryPort = strconv.Itoa(p.DockerRegistry.Port)
+	}
+	if p.DockerRegistry.Address != "" { // Use external registry
 		cc.DockerCAPath = p.DockerRegistry.CAPath
 		cc.DockerRegistryAddress = p.DockerRegistry.Address
-		cc.DockerRegistryPort = strconv.Itoa(p.DockerRegistry.Port)
-	} else if p.DockerRegistry.SetupInternal {
+	} else if p.DockerRegistry.SetupInternal { // Use registry on master[0]
 		cc.EnableInternalDockerRegistry = true
```
Comment (alexbrand) [inline pkg/install/execute.go]: It's somewhat unclear what the difference between `EnablePrivateDockerRegistry` and `EnableInternalDockerRegistry` is. Do we need both flags? If so, does `DeployInternalDockerRegistry` make more sense?

File: pkg/install/execute.go [hunk 3]
```diff
@@ -681,21 +680,17 @@ func (ae *ansibleExecutor) buildClusterCatalog(p *Plan) (*ansible.ClusterCatalog
 		cc.LoadBalancedFQDN = p.Master.Nodes[0].InternalIP
 	}
 
-	if p.DockerRegistry.Address != "" {
-		cc.EnableInternalDockerRegistry = false
+	if p.ConfigureDockerRegistry() {
 		cc.EnablePrivateDockerRegistry = true
+		cc.DockerRegistryAddress = p.DockerRegistryAddress()
+		cc.DockerCAPath = filepath.Join(tlsDir, "ca.pem")
+		cc.DockerRegistryPort = strconv.Itoa(p.DockerRegistry.Port)
```
Comment (alexbrand) [inline pkg/install/execute.go]: Do we respect this port when deploying the internal registry? Seems like this is changing behavior from the old default (`8443`)

File: pkg/install/plan_types.go [hunk 4]
```diff
@@ -289,3 +298,12 @@ func hasIP(nodes *[]Node, ip string) bool {
 func (p Plan) ConfigureDockerRegistry() bool {
 	return p.DockerRegistry.Address != "" || p.DockerRegistry.SetupInternal
 }
+
+func (p Plan) DockerRegistryAddress() string {
+	address := p.DockerRegistry.Address
+	// If external is not set usem master[0]
+	if address == "" {
+		address = p.Master.Nodes[0].InternalIP
```
Comment (alexbrand) [inline pkg/install/plan_types.go]: `InternalIP` can be empty

File: pkg/install/execute.go [hunk 5]
```diff
@@ -751,6 +746,15 @@ func (ae *ansibleExecutor) generateTLSAssets(p *Plan) error {
 	if err != nil {
 		return fmt.Errorf("error generating certificates for the cluster: %v", err)
 	}
+
+	// Generate kubeconfig
+	err = GenerateKubeconfig(p, ae.options.GeneratedAssetsDirectory)
```
Comment (alexbrand) [inline pkg/install/execute.go]: This seems unrelated to TLS assets. Was there a reason for moving this into here?

In general it seems like a good idea to move the kubeconfig generation so that it happens before the installation. I think we can achieve this by moving the call to this func higher up in `cli/apply.go`

File: pkg/install/helm.go [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+	Binary          string
+	ClientDirectory string
+	TillerImage     string
+	ServiceAccount  string
+	Kubeconfig      string
+	Stdout          io.Writer
+	Stderr          io.Writer
+}
+
+func DefaultHelmClient() (*HelmConfig, error) {
```
Comment (alexbrand) [inline pkg/install/helm.go]: Seems like this should return a `HelmClient`?

File: pkg/install/helm.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	home, err := homedir.Dir()
+	if err != nil {
+		return nil, fmt.Errorf("Could not determing $HOME directory: %v", err)
```
Comment (alexbrand) [inline pkg/install/helm.go]: typo: determine

File: pkg/install/helm.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	_, err := os.Stat(h.ClientDirectory)
+	var backedup bool
+	// Directory exists
+	if err == nil {
+		if err = os.Rename(h.ClientDirectory, h.BackupDirectory()); err != nil {
+			return true, fmt.Errorf("Could not back up %q directory: %v", h.ClientDirectory, err)
+		}
+		return true, nil
+	} else if !os.IsNotExist(err) { // Directory does not exist but got some other error
+		return backedup, fmt.Errorf("Could not determing if $HOME/.helm directory exists: %v", err)
```
Comment (alexbrand) [inline pkg/install/helm.go]: typo: determine

File: pkg/cli/apply.go [hunk 9]
```diff
@@ -88,34 +89,62 @@ func (c *applyCmd) run() error {
...
+		util.PrintHeader(c.out, "Installing Helm on the Cluster", '=')
+		helm, err := install.DefaultHelmClient()
+		if err != nil {
+			return fmt.Errorf("error getting a valid Helm client: %v", err)
+		}
+		helm.Kubeconfig = filepath.Join(c.generatedAssetsDir, "kubeconfig")
+		// On a disconnected install set tiller image with the correct tag
+		// Prepend the custom registry address and port
+		if plan.ConfigureDockerRegistry() && plan.Cluster.DisconnectedInstallation {
+			// TODO how to cleanly get the tiller tag?
```
Comment (alexbrand) [inline pkg/cli/apply.go]: Is this still a TODO?

File: pkg/install/helm.go [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+	}
+	// Directory does not already exist, nothing to do
+	return backedup, nil
+}
+
+// Init exectues 'helm init' using the local binary
+func (h *HelmConfig) Init() error {
```
Comment (alexbrand) [inline pkg/install/helm.go]: This is minor, but what do you think of this method taking the tiller image and kubeconfig as params. That way we don't have to set them on the struct.

File: pkg/cli/upgrade.go [hunk 11]
```diff
@@ -178,7 +178,7 @@ func doUpgrade(out io.Writer, opts *upgradeOpts) error {
 		fmt.Fprintln(out)
 	}
 
-	if plan.ConfigureDockerRegistry() && plan.Cluster.DisconnectedInstallation {
+	if plan.ConfgiureDockerWithPrivateRegistry() && plan.Cluster.DisconnectedInstallation {
```
Comment (alexbrand) [inline pkg/cli/upgrade.go]: Typo

File: pkg/install/plan_types.go [hunk 12]
```diff
@@ -285,7 +295,29 @@ func hasIP(nodes *[]Node, ip string) bool {
...
+		if p.Master.Nodes[0].InternalIP != "" {
+			address = p.Master.Nodes[0].InternalIP
+		}
+	}
+	return address
+}
+
+func (p Plan) DockerRegistryPort() string {
+	port := 8443
+	// If external is not set usem master[0]
```
Comment (alexbrand) [inline pkg/install/plan_types.go]: Comment seems irrelevant to code

File: pkg/cli/apply.go [hunk 13]
```diff
@@ -88,34 +91,61 @@ func (c *applyCmd) run() error {
 	if err != nil {
 		return fmt.Errorf("error validating plan: %v", err)
 	}
-	plan, err := c.planner.Read()
+	plan, _ := c.planner.Read()
```
Comment (alexbrand) [inline pkg/cli/apply.go]: Chances of this failing are slim b/c we've read the plan already in validate, but I'd rather handle this even if it's redundant. If this code moves around, or we change how we do validation, we'll most likely forget about this.

File: pkg/util/file.go [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+// BackupDirectory checks for existance of the $sourceDir and backs it up to backupDir
+func BackupDirectory(sourceDir string, backupDir string) (bool, error) {
+	_, err := os.Stat(sourceDir)
+	var backedup bool
+	// Directory exists
+	if err == nil {
+		if err = os.Rename(sourceDir, backupDir); err != nil {
+			return true, fmt.Errorf("Could not back up %q directory: %v", sourceDir, err)
```
Comment (alexbrand) [inline pkg/util/file.go]: Should this be `false` as nothing was backed up?

**Comment by dkoshkin** [comment]
Need to investigate further https://github.com/kubernetes/charts/issues/1014

**Comment by alexbrand** [comment]
This LGTM
